File: src/index/dense.ts
/** Uncompressed float32 semantic index. Debug/precision baseline only; production uses int8. */
import type { SemanticIndex } from "./semantic-index.ts";
import { TopKDistanceCollector } from "./topk.ts";

function cosineDistanceFlat(
  data: Float32Array,
  dim: number,
  docIndex: number,
  query: Float32Array,
): number {
  const offset = docIndex * dim;
  let dot = 0;
  for (let i = 0; i < dim; i++) {
    dot += (query[i] ?? 0) * (data[offset + i] ?? 0);
  }
  return 1 - dot;
}

export class VectorIndex implements SemanticIndex {
  private readonly data: Float32Array;
  private readonly count: number;
  private readonly dim: number;
  private vectorsCache: Float32Array[] | null = null;

  constructor(vectors: Float32Array[]) {
    if (vectors.length === 0) {
      this.count = 0;
      this.dim = 0;
      this.data = new Float32Array(0);
      return;
    }
    this.count = vectors.length;
    this.dim = vectors[0]?.length ?? 0;
    this.data = new Float32Array(this.count * this.dim);
    for (let i = 0; i < this.count; i++) {
      const vec = vectors[i];
      if (vec) {
        this.data.set(vec, i * this.dim);
      }
    }
  }

  static fromFlatBuffer(data: Float32Array, count: number, dim: number): VectorIndex {
    const index = Object.create(VectorIndex.prototype) as VectorIndex;
    Object.assign(index, { data, count, dim });
    return index;
  }

  get size(): number {
    return this.count;
  }

  get dimensions(): number {
    return this.dim;
  }

  getVectors(): readonly Float32Array[] {
    if (this.vectorsCache) {
      return this.vectorsCache;
    }
    const vectors: Float32Array[] = [];
    for (let i = 0; i < this.count; i++) {
      vectors.push(this.data.subarray(i * this.dim, (i + 1) * this.dim));
    }
    this.vectorsCache = vectors;
    return vectors;
  }

  vectorAt(docIndex: number): Float32Array {
    if (docIndex < 0 || docIndex >= this.count) {
      throw new Error(`Missing vector at index ${docIndex}`);
    }
    return this.data.subarray(docIndex * this.dim, (docIndex + 1) * this.dim);
  }

  memoryBytes(): number {
    return this.count * this.dim * 4;
  }

  query(
    queryVector: Float32Array,
    k: number,
    selector?: readonly number[],
  ): { indices: number[]; distances: number[] } {
    if (k < 1) {
      throw new Error(`k should be >= 1, is now ${k}`);
    }
    if (this.count === 0) {
      return { indices: [], distances: [] };
    }

    const effectiveK = Math.min(k, selector?.length ?? this.count);
    if (effectiveK === 0) {
      return { indices: [], distances: [] };
    }

    const collector = new TopKDistanceCollector(effectiveK);
    if (selector) {
      for (const idx of selector) {
        if (idx < 0 || idx >= this.count) {
          continue;
        }
        collector.offer(idx, cosineDistanceFlat(this.data, this.dim, idx, queryVector));
      }
    } else {
      for (let i = 0; i < this.count; i++) {
        collector.offer(i, cosineDistanceFlat(this.data, this.dim, i, queryVector));
      }
    }

    const top = collector.finish();
    return {
      indices: top.map((t) => t.index),
      distances: top.map((t) => t.distance),
    };
  }

  async save(dir: string): Promise<void> {
    await Bun.write(`${dir}/vectors.bin`, this.data);
    await Bun.write(
      `${dir}/meta.json`,
      JSON.stringify({ count: this.count, dimensions: this.dim, storage: "float32" }),
    );
  }

  static async load(dir: string): Promise<VectorIndex> {
    const meta = JSON.parse(await Bun.file(`${dir}/meta.json`).text()) as {
      count: number;
      dimensions: number;
    };
    const data = new Float32Array(await Bun.file(`${dir}/vectors.bin`).arrayBuffer());
    return VectorIndex.fromFlatBuffer(data, meta.count, meta.dimensions);
  }
}

File: internal/index/dense.go
package index

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
)

type QueryResult struct {
	Indices   []int
	Distances []float64
}

type VectorIndex struct {
	vectors [][]float32
}

func NewVectorIndex(vectors [][]float32) *VectorIndex {
	return &VectorIndex{vectors: vectors}
}

func (v *VectorIndex) Size() int {
	return len(v.vectors)
}

func (v *VectorIndex) Dimensions() int {
	if len(v.vectors) == 0 {
		return 0
	}
	return len(v.vectors[0])
}

func (v *VectorIndex) Vectors() [][]float32 {
	return v.vectors
}

func (v *VectorIndex) VectorAt(docIndex int) ([]float32, error) {
	if docIndex < 0 || docIndex >= len(v.vectors) {
		return nil, fmt.Errorf("missing vector at index %d", docIndex)
	}
	return v.vectors[docIndex], nil
}

func (v *VectorIndex) MemoryBytes() int {
	return len(v.vectors) * v.Dimensions() * 4
}

func (v *VectorIndex) Query(queryVector []float32, k int, selector ...[]int) (QueryResult, error) {
	if k < 1 {
		return QueryResult{}, fmt.Errorf("k should be >= 1, is now %d", k)
	}
	if len(v.vectors) == 0 {
		return QueryResult{}, nil
	}

	indices := make([]int, len(v.vectors))
	for i := range indices {
		indices[i] = i
	}
	if len(selector) > 0 && selector[0] != nil {
		indices = selector[0]
	}
	effectiveK := min(k, len(indices))
	if effectiveK == 0 {
		return QueryResult{}, nil
	}

	entries := make([]TopKDistanceEntry, 0, len(indices))
	for _, idx := range indices {
		if idx < 0 || idx >= len(v.vectors) {
			continue
		}
		entries = append(entries, TopKDistanceEntry{Index: idx, Distance: cosineDistance(queryVector, v.vectors[idx])})
	}
	top := SelectTopKByDistance(entries, effectiveK)
	return queryResultFromTop(top), nil
}

func (v *VectorIndex) Save(dir string) error {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}

	dim := v.Dimensions()
	count := v.Size()
	buf := make([]byte, count*dim*4)
	offset := 0
	for _, vector := range v.vectors {
		for _, value := range vector {
			binary.LittleEndian.PutUint32(buf[offset:], math.Float32bits(value))
			offset += 4
		}
	}
	if err := os.WriteFile(filepath.Join(dir, "vectors.bin"), buf, 0o644); err != nil {
		return err
	}
	return writeMeta(dir, semanticMeta{Count: count, Dimensions: dim, Storage: "float32"})
}

func LoadVectorIndex(dir string) (*VectorIndex, error) {
	meta, err := readMeta(dir)
	if err != nil {
		return nil, err
	}
	raw, err := os.ReadFile(filepath.Join(dir, "vectors.bin"))
	if err != nil {
		return nil, err
	}
	vectors := make([][]float32, meta.Count)
	for i := 0; i < meta.Count; i++ {
		vector := make([]float32, meta.Dimensions)
		for j := 0; j < meta.Dimensions; j++ {
			offset := (i*meta.Dimensions + j) * 4
			vector[j] = math.Float32frombits(binary.LittleEndian.Uint32(raw[offset:]))
		}
		vectors[i] = vector
	}
	return NewVectorIndex(vectors), nil
}

func cosineDistance(a, b []float32) float64 {
	var dot float64
	for i := range a {
		var bv float32
		if i < len(b) {
			bv = b[i]
		}
		dot += float64(a[i]) * float64(bv)
	}
	return 1 - dot
}

func queryResultFromTop(top []TopKDistanceEntry) QueryResult {
	result := QueryResult{
		Indices:   make([]int, len(top)),
		Distances: make([]float64, len(top)),
	}
	for i, entry := range top {
		result.Indices[i] = entry.Index
		result.Distances[i] = entry.Distance
	}
	return result
}

type semanticMeta struct {
	Count      int    `json:"count"`
	Dimensions int    `json:"dimensions"`
	Storage    string `json:"storage,omitempty"`
}

func writeMeta(dir string, meta semanticMeta) error {
	data, err := json.Marshal(meta)
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, "meta.json"), data, 0o644)
}

func readMeta(dir string) (semanticMeta, error) {
	var meta semanticMeta
	data, err := os.ReadFile(filepath.Join(dir, "meta.json"))
	if err != nil {
		return meta, err
	}
	err = json.Unmarshal(data, &meta)
	return meta, err
}
